Add an implementation of alias method for weighted indices

zroug opened the pull request:
This is the pull request that was requested in #601. It adds an alternative implementation for `WeightedIndex` that is based on the alias method which has `O(1)` sampling speed.

In comparison to the code I posted in #601 I changed the error handling to better match the existing `WeightedIndex` implementation and improved sampling performance.

I'm not sure on the API though. Having this algorithm only for `f64` weights seems a bit restricting. Especially when I consider that for the use case I originally created this implementation for I used `f32` weights to reduce memory usage. While this algorithm can be adapted to work with integer weights, a multiplication with a `usize` (the number of weights) would still be necessary which would make that potentially prone to overflows.

dhardy commented:
@huonw review?

huonw commented:
I apologise, I'm not in a position to review PRs to `rand` at the moment; hopefully someone else can take up the task. (A brief glance over the code seems reasonable, but I haven't dug into it to understand it deeply.)

vks commented:
* Doc comments are still missing. I presume this is the reason for the [WIP].
* To be able to replace `WeightedIndex`, it needs to be implemented for weights that implement `SampleUniform + PartialOrd` instead of just `f64`.
* Here are the benchmark results with my machine:
```
test distr_weighted_alias_method           ... bench:      10,912 ns/iter (+/- 351) = 733 MB/s
test distr_weighted_alias_method_large_set ... bench:      12,174 ns/iter (+/- 415) = 657 MB/s
test distr_weighted_f64                    ... bench:       7,745 ns/iter (+/- 374) = 1032 MB/s
test distr_weighted_i8                     ... bench:      10,587 ns/iter (+/- 287) = 755 MB/s
test distr_weighted_large_set              ... bench:      63,551 ns/iter (+/- 2,587) = 125 MB/s
test distr_weighted_u32                    ... bench:      10,874 ns/iter (+/- 536) = 735 MB/s
```
So the new method seems to be slower for small sets, but much faster for large sets.

zroug commented:
I made the implementation generic. I created a custom trait for now but that could be changed when rust has more traits built in. I also tried to keep checking for errors for types that behave like primitive types. For the benchmarks I have used the exact same ones, that the existing implementation uses.

zroug commented on src/distributions/weighted.rs:
```diff
@@ -130,6 +130,165 @@ impl<X> Distribution<usize> for WeightedIndex<X> where
+
+        // Prepare a distribution to sample random indices. Creating it beforehand
+        // improves sampling performance.
+        let uniform_index = super::Uniform::new(0, no_alias_odds.len());
```
I usually do that when there are multiple collections with a common length, so you don't have to arbitrarily pick one to take the length from. I have fixed it here.

vks commented on src/distributions/weighted.rs:
```diff
@@ -131,40 +133,52 @@ impl<X> Distribution<usize> for WeightedIndex<X> where
 #[allow(missing_docs)] // todo: add docs
-#[derive(Debug, Clone)]
-pub struct AliasMethodWeightedIndex {
+#[allow(missing_debug_implementations)] // todo: why does `#[derive(Debug)]` not work?
```
You probably would have to require `W` to implement `Debug`.

zroug replied:
I tried that but it didn't work.

dhardy replied:
It would probably work if you force `trait UniformSampler: Debug + Sized`, except probably other stuff would break... try:
```
where W: Debug, super::Uniform<W>: Debug
```

dhardy replied:
What were you trying to do though? Use the `derive(Debug)` macro? That [cannot currently add the correct bounds](https://github.com/rust-lang/rust/issues/26925). I think to implement `Debug` correctly here you would need to do so manually (with the above `where` clause).

zroug replied:
Ah, I wasn't aware of https://github.com/rust-lang/rust/issues/26925. I have implemented `Debug` and `Clone` myself.

dhardy commented on src/distributions/weighted.rs:
```diff
@@ -298,6 +297,38 @@ impl<W: AliasMethodWeight> Distribution<usize> for AliasMethodWeightedIndex<W> {
+impl<W: AliasMethodWeight> ::core::fmt::Debug for AliasMethodWeightedIndex<W>
+where
+    W: ::core::fmt::Debug,
+    super::Uniform<W>: ::core::fmt::Debug,
```
You can `use core::fmt` already, and `Result` is already in the prelude.

zroug replied:
Fixed.

zroug commented:
Regarding the documentation: Can I copy relevant parts from the existing documentation? Also English isn't my native langue, so you may have to correct some things when I write the documentation.

dhardy commented:
Yes, you can copy what you need. Sure, I'll be happy to review for English errors, but you seem to know the language quite well already!

zroug commented:
I removed the [WIP] because I have added everything I wanted to add. Of course I will still address any review issues. Should I squash the commits before merge?

dhardy commented:
We don't require squashing commits, though if there are many small ones it can be nice to tidy them up a little.

zroug commented:
I cleaned up the commits a bit.

vks commented:
Updated benchmarks:
```
test distr_weighted_alias_method_f64       ... bench:      10,417 ns/iter (+/- 238) = 767 MB/s
test distr_weighted_alias_method_i8        ... bench:       9,937 ns/iter (+/- 364) = 805 MB/s
test distr_weighted_alias_method_large_set ... bench:      11,027 ns/iter (+/- 299) = 725 MB/s
test distr_weighted_alias_method_u32       ... bench:       9,904 ns/iter (+/- 336) = 807 MB/s
test distr_weighted_f64                    ... bench:       8,598 ns/iter (+/- 289) = 930 MB/s
test distr_weighted_i8                     ... bench:      10,597 ns/iter (+/- 432) = 754 MB/s
test distr_weighted_large_set              ... bench:      63,551 ns/iter (+/- 4,722) = 125 MB/s
test distr_weighted_u32                    ... bench:      11,003 ns/iter (+/- 561) = 727 MB/s
```

So the alias method seems to be faster for integer and large float sets, but a bit slower for small float sets. I think it might make sense to drop our existing implementation.

vks commented on src/distributions/weighted.rs:
```diff
@@ -130,6 +132,342 @@ impl<X> Distribution<usize> for WeightedIndex<X> where
+        let (a, b) = values.split_at(mid);
+        pairwise_sum(a) + pairwise_sum(b)
+    }
+}
```
Instead of the pairwise sum, [an algorithm proven to be correct](http://www.cs.cmu.edu/~quake-papers/robust-arithmetic.ps) could be used. This is [internally used in libtest](https://github.com/rust-lang/rust/blob/31eb0e2d3c0bfd2e0f5b662d82971c1a203708be/src/libtest/stats.rs#L154). For integers, the naive summation is fine. It might be more efficient to move the implementation of the numerically correct summation to `AliasMethodWeight`, so it can be specialized for integers.

zroug replied:
The sum does not need to be exact because an approximated sum only reduces the accuracy of the sampling and doesn't make it completely wrong. I just wanted to avoid the effect that the sum doesn't change at all when there are many weights, which for `f32` weights happens pretty early: https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=9b2e8d44c8e4dc780368ecf4791764b1. Pairwise summation, which does have an upper bound on the error too (https://en.wikipedia.org/wiki/Pairwise_summation), avoids that and at the same time doesn't need any more addition operations than naive summation.

In 8641a9bb1624e9a984a69ae0c24fcf16ee846bc8 I changed the code, so that pairwise summation is only used for floating point weights.

vks replied:
Fair enough, this seems like a reasonable tradeoff for performance.

vks commented on src/distributions/weighted.rs:
```diff
@@ -130,6 +132,342 @@ impl<X> Distribution<usize> for WeightedIndex<X> where
+/// Sampling an [`AliasMethodWeightedIndex<W>`] distribution returns the index
+/// of a randomly selected element from the vector used to create the
+/// [`AliasMethodWeightedIndex<W>`]. The chance of a given element being picked
+/// is proportional to the value of the element. The weights can have any  type
```
There is a redundant whitespace in front of "type".

zroug replied:
Fixed in 9c44b6aa510b25408ce050d2785b5a9fcb4e98de.

vks commented on src/distributions/weighted.rs:
```diff
@@ -130,6 +132,342 @@ impl<X> Distribution<usize> for WeightedIndex<X> where
+///
+/// Given that `n` is the number of items in the vector used to create an
+/// [`AliasMethodWeightedIndex<W>`], [`AliasMethodWeightedIndex<W>`] will take
+/// up `O(n)` amount of memory. More specifically it takes up some constant
```
Should probably be "up to" or "will require".

zroug replied:
Fixed in 9c44b6aa510b25408ce050d2785b5a9fcb4e98de.

vks commented on src/distributions/weighted.rs:
```diff
@@ -130,6 +132,342 @@ impl<X> Distribution<usize> for WeightedIndex<X> where
+    }
+}
+
+/// In comparision to naive accumulation, the pairwise sum algorithm reduces
```
Should be "comparison".

zroug replied:
Fixed in 9c44b6aa510b25408ce050d2785b5a9fcb4e98de.

vks commented on src/distributions/weighted.rs:
```diff
@@ -130,6 +132,342 @@ impl<X> Distribution<usize> for WeightedIndex<X> where
+    /// Element of `Self` equivalent to 0.
+    const ZERO: Self;
+
+    /// Converts an [`usize`] to a `Self`, rounding if necessary.
```
Should be "a usize".

zroug replied:
Fixed in 9c44b6aa510b25408ce050d2785b5a9fcb4e98de.

dhardy commented on src/distributions/weighted.rs:
```diff
@@ -228,3 +698,40 @@ impl fmt::Display for WeightedError {
+
+/// Error type returned from [`AliasMethodWeightedIndex::new`].
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum AliasMethodWeightedIndexError {
```
Is there much justification for having a separate error type? It adds a significant amount of code for little gain.

I don't have a problem with modifying the members of `WeightedError`, though of course such a change can't land in a patch release.

zroug replied:
I don't think so. In the first version of the pull request the error types were a little bit more different. Now they are pretty similar. Let me know if you want to make that breaking change.

dhardy replied:
Yes, I would prefer to do it this way. I won't be able to merge yet with a breaking change, but I'm not sure there will be more releases for 0.6.x anyway so it likely won't make much difference to when it gets released.

vks replied:
Didn't we already merge some breaking change to master?

zroug replied:
I will change this once there is a plan for the other discussion, because it might affect this.

zroug replied:
I changed this in 5b2934122f04e1ff41024b3c0597ace711d21f08.

dhardy commented on src/distributions/mod.rs:
```diff
@@ -181,7 +181,11 @@ pub use self::other::Alphanumeric;
+#[cfg(feature = "alloc")]
+pub use self::weighted::{
+    AliasMethodWeight, AliasMethodWeightedIndex, AliasMethodWeightedIndexError, WeightedError,
+    WeightedIndex,
```
Alternatively we could make the module public and name these `weighted::{AliasMethodWeight, AliasMethod, Error, BinarySearch}`. Thoughts?

(obviously a breaking change)

zroug replied:
I think that would be better and more like it is done in the standard library  (`std::io::Error`). Again, let me know if you want to do that.

zroug replied:
But maybe the module must be renamed because with the proposed naming the 'index' part is lost.

dhardy replied:
I'm not too sure we need to keep `Index` in the name anyway; it's the only type of weighted sampling we have. It's not the best idea to break this stuff *again*, but still better to get it right than leave a mess IMO.

But lets wait for @vks to comment.

vks replied:
As I mentioned above, I think we might get rid of our current implementation and only have the alias method. Then there wouldn't be naming issues.

Alternatively, if we decide to keep both, I would prefer to drop the common `AliasMethod` prefix and instead have an `alias_method` module. This would make it much easier for users to switch between the two implementations.

dhardy replied:
The current benchmarks show the alias method to always be in the lead or only slightly behind, *however* if you move the set-up time into the measurement loop, then the binary-search method can be significantly faster (three times faster on the large_set bench with 1000 samples; nine times faster with 100 samples, and a little faster on the smaller sets).

Memory usage will be a little higher with the Alias method due to the extra `Vec<usize>`; mostly this is unimportant I think (unless memory constrained and having a large set of weights in a small type).

The Alias method has some extra requirements on the type, notably `Copy`. Should we use `Clone` instead?

---

I think there is room for both implementations, though the current presentation and documentation is not ideal. So what do you think about the following structure?
```rust
distributions::{
    weighted::{
        alias::{WeightedIndex, Weight},
        WeightedIndex, Error
    },
}
```

vks replied:
> The Alias method has some extra requirements on the type, notably Copy. Should we use Clone instead?

I missed that. In that case we should probably have both, and working with `Clone` would be nicer.

> So what do you think about the following structure?

This is what I would suggest as well.

dhardy replied:
Good. @zroug would you make these changes please?

zroug replied:
Yes, I will make these changes.

zroug replied:
I did these changes in 5b2934122f04e1ff41024b3c0597ace711d21f08 but I used `alias_method` instead of `alias` as module name. Just `alias` didn't sound right to me and the word alias has a much broader meaning. Are you okay with this?

I wasn't sure if I should keep the reexports for `WeightedIndex`. I have kept them.

dhardy commented:
There are still a few broken links in the documentation, and it would be nice to have some comparative information in the `weighted` module doc, but these can come later, so I think it's time to merge this.

One caveat, as @zroug mentions, is that `distributions::WeightedIndex` is now an alias for `distributions::weighted::WeightedIndex`. The best option may be to deprecate the re-exports, *however* we should probably hold off on this for now since we might instead move this to another crate like `rand_distr::weighted::*`.

@vks do you agree? If so we can merge this as-is then prepare some doc link fixes.

vks commented:
@dhardy Yes, I agree!

> The best option may be to deprecate the re-exports, however we should probably hold off on this for now since we might instead move this to another crate like rand_distr::weighted::*.

Note that `rand::seq` depends on `WeightedIndex`. Should this also be moved to a different crate?

dhardy commented:
Good point. Maybe it is best just to deprecate `distributions::WeightedIndex` then?
